### PR TITLE
Fix key lookup

### PR DESCRIPTION
**Pull request by ahgittin**
fixes subtle bug which prevents some evaluation strategies for map subkeys

### CODE REVIEW

**Comment by ahgittin** [comment]
BTW i'm not enamoured of the ambiguity in map config keys dot syntax ... it was useful in the properties file days but not anymore imo ... but we do need to be careful and deprecate it, not just chop it off at the knees which the change to config inheritance lookup did inadvertently

File: camp/camp-brooklyn/src/test/resources/org/apache/brooklyn/camp/brooklyn/config-nested-test.bom [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+      brooklyn.config:
+        template.substitutions:
+          field: val
+  - id: base-map-parameter
+    itemType: entity
+    item:
+      type: org.apache.brooklyn.entity.stock.BasicEntity
+      brooklyn.parameters:
+        - name: template.substitutions
+          type: java.util.Map
```
Comment (neykov) [inline camp/camp-brooklyn/src/test/resources/org/apache/brooklyn/camp/brooklyn/config-nested-test.bom]: We should probably create a `MapConfigKey` if we see this in parameters.

File: core/src/main/java/org/apache/brooklyn/core/config/internal/AbstractConfigMapImpl.java [hunk 2]
```diff
@@ -356,8 +357,9 @@ protected ConfigInheritance getDefaultRuntimeInheritance() {
             
             Function<TContainer, Maybe<Object>> lookupFn = new Function<TContainer, Maybe<Object>>() {
                 @Override public Maybe<Object> apply(TContainer input) {
-                    Maybe<Object> result = getRawValueAtContainer(input, queryKey);
-                    if (!raw) result = resolveRawValueFromContainer(input, queryKey, result);
+                    // lookup against ownKey as it may do extra resolution (eg grab *.* subkeys if a map)
```
Comment (neykov) [inline core/src/main/java/org/apache/brooklyn/core/config/internal/AbstractConfigMapImpl.java]: Makes sense. Also good testo coverage. Merging.

File: camp/camp-brooklyn/src/test/java/org/apache/brooklyn/camp/brooklyn/ConfigNestedYamlTest.java [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+        Entity app = createAndStartApplication( bp );
+        waitForApplicationTasks(app);
+        return Iterables.getOnlyElement( app.getChildren() );
+    }
+    
+    protected Entity checkEntity(Entity ent, boolean expectDotSyntaxToWork) throws Exception {
+        String dotSyntax = ent.getConfig( ConfigKeys.newStringConfigKey("template.substitutions.field") );
+        Assert.assertEquals(dotSyntax, expectDotSyntaxToWork ? "val" : null, "wrong value for dot syntax: "+dotSyntax);
+        // observed in some situations we get the dot syntax instead!
+        // Assert.assertEquals(dotSyntax, "val", "wrong value for dot syntax: "+dotSyntax);
```
Comment (neykov) [inline camp/camp-brooklyn/src/test/java/org/apache/brooklyn/camp/brooklyn/ConfigNestedYamlTest.java]: I'd have expected this to always be the case, for any `Map` value, not just for `MapConfigKey`. The logic doing the lookup (if I remember correctly) is trying to get for increasingly longer parts of the full key until it has a hit.

**Pull request merged**